### ISSUE TITLE
[REVIEW]: City2BA: Tools for creating synthetic bundle adjustment problems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3125]
**Submitting author:** @tkonolige (<a href="http://orcid.org/0000-0002-5052-6479">Tristan Konolige</a>)
**Repository:** <a href="https://github.com/tkonolige/city2ba" target ="_blank">https://github.com/tkonolige/city2ba</a>
**Version:** v1.1.0
**Editor:** @poulson
**Reviewer:** @adi3, @breezykermo
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59/status.svg)](https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59"><img src="https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59/status.svg)](https://joss.theoj.org/papers/45fcd655d6c229b5673c8c16781d8b59)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/tkonolige/city2ba" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@adi3 & @breezykermo, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @poulson know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/tkonolige/city2ba">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@tkonolige) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/tkonolige/city2ba">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@tkonolige) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @adi3, @breezykermo it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Failed to discover a `Statement of need` section in paper

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.03 s (445.4 files/s, 94731.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Rust                             8            298            376           2294
Markdown                         2             18              0             82
TOML                             1              2              0             34
TeX                              1              1              0             31
YAML                             1              2              0             26
JSON                             1              0              0             23
Dockerfile                       1              0              0              3
-------------------------------------------------------------------------------
SUM:                            15            321            376           2493
-------------------------------------------------------------------------------


Statistical information for the repository '56bc84eb0646492fd6cfbb13' was
gathered on 2021/03/22.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- 10.1007/978-3-319-10578-9_5 may be a valid DOI for title: Robust global translations with 1DSFM
- 10.1109/iccv.2013.122 may be a valid DOI for title: Street view motion-from-structure-from-motion

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03125/joss.03125/10.21105.joss.03125.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03125/joss.03125/10.21105.joss.03125.pdf) :page_facing_up: :point_left:

**Comment by arfon** [comment]
@adi3, @breezykermo - Thanks again for agreeing to review this submission! This is the review thread for the paper. All of our communications will happen here from now on.

Please read the "Reviewer instructions & questions" in the first comment above.

Both reviewers have checklists at the top of this thread (in that first comment) with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention https://github.com/openjournals/joss-reviews/issues/3125 so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for the review process to be completed within about 4-6 weeks but please make a start well ahead of this as JOSS reviews are by their nature iterative and any early feedback you may be able to provide to the author will be very helpful in meeting this schedule.

**Comment by adi3** [comment]
@tkonolige can you please fix the missing DOIs in your paper listed above?

**Comment by whedon** [comment]
:wave: @adi3, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @breezykermo, please update us on how your review is going (this is an automated reminder).

**Comment by adi3** [comment]
Waiting on input from @tkonolige to proceed.

**Comment by tkonolige** [comment]
@adi3 DOIs have been added.

**Comment by adi3** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by adi3** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1007/978-3-319-10578-9_5 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1109/iccv.2013.122 is INVALID because of 'https://doi.org/' prefix
```

**Comment by adi3** [comment]
@tkonolige Please fix invalid DOIs and then run `@whendon check references` to make sure your DOIs list as OK.

**Comment by tkonolige** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/978-3-319-10578-9_5 is OK
- 10.1109/iccv.2013.122 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by arfon** [comment]
:wave: @adi3, @breezykermo – just checking in here. It looks like @adi3 has made some progress on their review but @breezykermo doesn't look to have started yet.

@breezykermo - are you still able to complete this review for us?

**Comment by adi3** [comment]
@tkonolige Could you please add some example usage of the tool in the repo's README? This will help us test functionality of the system and will also be useful to your end-users in getting started. Thanks!

**Comment by tkonolige** [comment]
@adi3 I have some example usage in the readme. Is there something missing from it that you would like you would like to see? Right now you have to provide your own .obj file, but I could have the example download one.

**Comment by arfon** [comment]
Hi all, just checking in on this review. It has been close to 5 months since the last update here.

@adi3, @breezykermo – could you give an update on your status here and ability to complete your reviews?

**Comment by arfon** [comment]
@whedon assign me as editor

**Comment by arfon** [comment]
Also, I'm going to take this submission on on behalf of @poulson.

**Comment by adi3** [comment]
@arfon this slipped through the nets as I was focused on my editorial duties. I'll get to this review sometime this month. Those automated whedon reminders are a boon to forgotten reviews.

**Comment by arfon** [comment]
> @arfon this slipped through the nets as I was focused on my editorial duties. I'll get to this review sometime this month. Those automated whedon reminders are a boon to forgotten reviews.

:zap: thanks @adi3!

**Comment by adi3** [comment]
@tkonolige could you please add an example `.obj` file in the repository so users can test the tool right out-of-the-box? Thanks!

**Comment by tkonolige** [comment]
@adi3 done

**Comment by adi3** [comment]
@tkonolige having trouble installing the tool. Cargo is existing with this error:

```
note: /usr/bin/ld: cannot find -lembree3
          collect2: error: ld returned 1 exit status
          

error: failed to compile `city2ba v1.1.0`, intermediate artifacts can be found at `/tmp/cargo-installKuVNju`

Caused by:
  could not compile `city2ba` due to previous error
```

I'm on Ubuntu LTS 20.04. Can you please see what might be wrong? Either you can fix this, or perhaps you could let me know which exact platform you tested your software on?

**Comment by arfon** [comment]
:wave: @adi3, @breezykermo – just checking in again here on progress with your reviews.

**Comment by adi3** [comment]
@tkonolige - here's my feedback so far:

- I installed `embree` on a fresh Linux box and adding the env var like you suggested indeed did the trick. I would recommend you add both these things to the installation instructions in a code block. Or at least the env export part.
- Is there a way to simply this and install `embree` directly via cargo?
- After installation, cargo asks the user to add the bin to their env `PATH`. Could you please add that step in the installation guide please? I believe it would be a simple `export PATH=$PATH:/home/ubuntu/.cargo/bin`.
- In the functionality testing, please add a quick line about downloading the `test_scene.obj`.
- It seems the program API has changed (`--num-cameras` to `--cameras`, `--num-points` to `--points`, etc). Can you please update the documentation to reflect these changes?
- I get this error when trying to run `generate` on the test_scene object: `Cannot generate world points with 0 cameras.` Can you please look into it and tell me the fixes? Here's the stacktrace:
```
> city2ba generate test_scene.obj problem.bal --cameras 100 --points 200
Generated 0 cameras
Modified intrinsics
thread 'main' panicked at 'Cannot generate world points with 0 cameras.', /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/city2ba-1.1.0/src/generate.rs:366:9
stack backtrace:
   0: std::panicking::begin_panic
   1: city2ba::generate::generate_world_points_uniform
   2: city2ba::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by adi3** [comment]
@tkonolige bumping this to the top of your inbox for when you have a chance!

**Comment by arfon** [comment]
:wave: @tkonolige – just checking in here to see if you've had a chance to look into @adi3's feedback?

**Comment by tkonolige** [comment]
@adi3 
> * I installed `embree` on a fresh Linux box and adding the env var like you suggested indeed did the trick. I would recommend you add both these things to the installation instructions in a code block. Or at least the env export part.
Added
> * Is there a way to simply this and install `embree` directly via cargo?
Yes, but it is not easy and would require work in an upstream dependency.
> * After installation, cargo asks the user to add the bin to their env `PATH`. Could you please add that step in the installation guide please? I believe it would be a simple `export PATH=$PATH:/home/ubuntu/.cargo/bin`.
Done
> * In the functionality testing, please add a quick line about downloading the `test_scene.obj`.
Done
> * It seems the program API has changed (`--num-cameras` to `--cameras`, `--num-points` to `--points`, etc). Can you please update the documentation to reflect these changes?
Done
> * I get this error when trying to run `generate` on the test_scene object: `Cannot generate world points with 0 cameras.` Can you please look into it and tell me the fixes? Here's the stacktrace:
Looks like not enough locations for cameras were generated, try increasing the number for `--cameras`. (I've added a statement to that effect to the error message).

**Comment by arfon** [comment]
@adi3 - could I ask you to take another look here to see if these updates meet your expectations?

**Comment by adi3** [comment]
@whedon generate pdf

**Comment by adi3** [comment]
@editorialbot generate pdf

**Comment by adi3** [comment]
@tkonolige - I'm almost finished reviewing this submission. Functionality-wise, I'm satisfied. However, in your paper, we need a section titled "Statement of Need" which will include most of the information you already have in your first paragraph. The Summary section itself can just be a few lines long and should provide only a high-level overview of the software and its benefits. Once you make these quick changes, I'm happy to check off the last few items.

**Comment by arfon** [comment]
@tkonolige – just checking you saw the most recent comment from @adi3. Did you manage to address this feedback?

**Comment by arfon** [comment]
👋 @tkonolige – just checking in again here. Can you confirm you're still interested in publishing this paper in JOSS?

**Comment by arfon** [comment]
(Just an FYI that I have emailed @tkonolige to clarify their plans here)

**Comment by tkonolige** [comment]
I've added a statement of need. Hopefully that covers all your concerns.

**Comment by arfon** [comment]
:wave: @adi3 – could you take a look at the Statement of Need and confirm this meets your expectations? Thanks!

**Comment by adi3** [comment]
@editorialbot commands

**Comment by adi3** [comment]
@editorialbot generate pdf

**Comment by adi3** [comment]
@tkonolige - thanks for adding the section. It reads fine. My only feedback would be that you put the *Statement of Need* section comes before the *Summary* section to make the paper flow better.

@arfon - This completes my checklist. Everything is ticked off now.

**Comment by arfon** [comment]
@tkonolige – I made some suggested edits to your paper – could you take a look at let me know what you think? https://github.com/tkonolige/city2ba/pull/5

**Comment by arfon** [comment]
Also @tkonolige – could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by arfon** [comment]
@editorialbot remove @breezykermo as reviewer

**Comment by tkonolige** [comment]
Here is the zenodo: https://zenodo.org/record/7523441

**Comment by arfon** [comment]
@editorialbot set 10.5281/zenodo.7523441 as archive

**Comment by arfon** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@adi3 – many thanks for your reviews here! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@tkonolige – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by tkonolige** [comment]
Thanks all!
